src/http/server.cr

```diff
@@ -95,90 +100,140 @@ class HTTP::Server
+  def self.new(handlers : Array(HTTP::Handler), &handler : HTTP::Handler::Proc) : self
+    new(HTTP::Server.build_middleware(handlers, handler))
   end
 
-  def self.new(port, handler)
-    new("127.0.0.1", port, handler)
+  # Creates a new HTTP server with the *handlers* array as handler chain.
+  def self.new(handlers : Array(HTTP::Handler)) : self
+    new(HTTP::Server.build_middleware(handlers))
   end
 
-  def initialize(@host : String, @port : Int32, &handler : Context ->)
+  # Creates a new HTTP server with the given *handler*.
+  def initialize(handler : HTTP::Handler | HTTP::Handler::Proc)
     @processor = RequestProcessor.new(handler)
   end
 
-  def initialize(@host : String, @port : Int32, handlers : Array(HTTP::Handler), &handler : Context ->)
-    handler = HTTP::Server.build_middleware handlers, handler
-    @processor = RequestProcessor.new(handler)
-  end
+  # Creates a `TCPServer` listenting on `host:port` and adds it as a socket, returning the local address
+  # and port the server listens on.
+  #
+  # If *reuse_port* is `true`, it enables the `SO_REUSEPORT` socket option,
+  # which allows multiple processes to bind to the same port.
+  def bind_tcp(host : String, port : Int32, reuse_port : Bool = false) : Socket::IPAddress
+    tcp_server = TCPServer.new(host, port, reuse_port: reuse_port)
 
-  def initialize(@host : String, @port : Int32, handlers : Array(HTTP::Handler))
-    handler = HTTP::Server.build_middleware handlers
-    @processor = RequestProcessor.new(handler)
+    bind(tcp_server)
+
+    tcp_server.local_address
   end
 
-  def initialize(@host : String, @port : Int32, handler : HTTP::Handler | HTTP::Handler::Proc)
-    @processor = RequestProcessor.new(handler)
+  # Creates a `TCPServer` listenting on `127.0.0.1:port` and adds it as a socket,
+  # returning the local address and port the server listens on.
+  #
+  # If *reuse_port* is `true`, it enables the `SO_REUSEPORT` socket option,
+  # which allows multiple processes to bind to the same port.
+  def bind_tcp(port : Int32, reuse_port : Bool = false) : Socket::IPAddress
+    bind_tcp "127.0.0.1", port, reuse_port
   end
 
-  # Returns the TCP port the server is connected to.
+  # Creates a `TCPServer` listening on an unused port and adds it as a socket.
+  #
+  # Returns the `Socket::IPAddress` with the determined port number.
   #
-  # For example you may let the system choose a port, then report it:
   # ```
-  # server = HTTP::Server.new(0) { }
-  # server.bind
-  # server.port # => 12345
+  # server = HTTP::Server.new { }
+  # server.bind_unused_port # => Socket::IPAddress.new("127.0.0.1", 12345)
   # ```
-  def port
-    if server = @server
-      server.local_address.port.to_i
-    else
-      @port
+  def bind_unused_port(host : String = "127.0.0.1", reuse_port : Bool = false) : Socket::IPAddress
+    bind_tcp host, 0, reuse_port
+  end
+
+  # Adds a `Socket::Server` *socket* to this server.
+  def bind(socket : Socket::Server) : Socket::Server
```

Comment:
Why return the server?

Reply:
It's intended as a simplification. 
This makes it easier to retrieve the actual address from binding a server whose local address depends on binding it: `http_server.bind(TCPServer.new(0)).address`

Reply:
But yeah, it's not such a clean API and can probably be removed.

Reply:
I vote remove it, variables work fine.